root/static/scripts/account/components/PreferencesForm.js

```diff
@@ -0,0 +1,256 @@
+/*
+ * @flow
+ * Copyright (C) 2018 MetaBrainz Foundation
+ *
+ * This file is part of MusicBrainz, the open internet music database,
+ * and is licensed under the GPL version 2, or (at your option) any
+ * later version: http://www.gnu.org/licenses/gpl-2.0.txt
+ */
+
+import * as React from 'react';
+import moment from 'moment';
+import 'moment-strftime';
+import 'moment-timezone';
+import noop from 'lodash/noop';
+
+import FormRow from '../../../../components/FormRow';
+import FormRowCheckbox from '../../../../components/FormRowCheckbox';
+import FormRowSelect from '../../../../components/FormRowSelect';
+import FormSubmit from '../../../../components/FormSubmit';
+import {l, N_l} from '../../common/i18n';
+import {Lens, prop, set, compose3} from '../../common/utility/lens';
+import hydrate from '../../../../utility/hydrate';
+
+type PreferencesFormT = FormT<{|
+  +datetime_format: FieldT<string>,
+  +email_on_no_vote: FieldT<boolean>,
+  +email_on_notes: FieldT<boolean>,
+  +email_on_vote: FieldT<boolean>,
+  +notify_via_email: FieldT<boolean>,
+  +public_ratings: FieldT<boolean>,
+  +public_subscriptions: FieldT<boolean>,
+  +public_tags: FieldT<boolean>,
+  +show_gravatar: FieldT<boolean>,
+  +subscribe_to_created_artists: FieldT<boolean>,
+  +subscribe_to_created_labels: FieldT<boolean>,
+  +subscribe_to_created_series: FieldT<boolean>,
+  +subscriptions_email_period: FieldT<string>,
+  +timezone: FieldT<string>,
+|}>;
+
+type Props = {|
+  +form: PreferencesFormT,
+|};
+
+type State = {|
+  form: PreferencesFormT,
+|};
+
+const timezoneOptions = {
+  grouped: false,
+  options: moment.tz.names().map(v => ({
```

Comment:
These options aren't in sync with what `DateTime::TimeZone` has. :(

https://gist.github.com/mwiencek/3261f94f4e4c6bb63eeb28348bd86535

So, many of those options (e.g. Africa/Addis_Ababa) will be rejected by the form. I guess `MusicBrainz::Server::Plugin::UserDate` won't work with them either since it uses `DateTime::TimeZone` under the hood.

Maybe updating both modules will solve this if the latest versions have the latest zoneinfo.

Reply:
I updated both Javascript `moment-timezone` and Perl `DateTime::TimeZone` to use the latest IANA TimeZone Database (2018e) with https://github.com/metabrainz/musicbrainz-server/pull/718/commits/30ae79cfe23699aa2dcbf1c3a89fc7683544567a. How to update `cpanfile.snapshot` from there?

It still doesn’t solve the issue as JS module includes timezone aliases while Perl module doesn’t. Forking either module is not desirable (even though it is probably possible for moment-timezone). The best path seems to be serializing options for the field timezone from Perl to JSON. This is exactly the same issue I addressed at https://github.com/metabrainz/musicbrainz-server/pull/730/commits/6bf052f9e07bc0125130410a409dcff7924a0638#diff-5053a22c07db9b62eceff446a9b99731R414. Shouldn’t options (or at least some set of valid values) be serialized for most of Select fields? Would it be fine to implement it from [Form::Role::ToJSON#L46](https://github.com/metabrainz/musicbrainz-server/blob/master/lib/MusicBrainz/Server/Form/Role/ToJSON.pm#L46)?

Reply:
Yeah, I agree we should serialize the options for the form somewhere. It might be tricky to do that in `Form::Role::ToJSON` in general though, because some select options require special customizations or have optgroups added (see `MusicBrainz::Server::Form::Utils`), so we'd need a way to customize the serialization for certain fields. Also, for repeatable fields the options are associated with the subfield, but we only want to serialize them once for the entire form. Just putting the options on the parent repeatable may not work if there are nested repeatables.

For now, maybe we should just serialize the options in the controller, using one of the subroutines from `MusicBrainz::Server::Form::Utils`, then pass them as a prop to the top-level form component. What do you think?

Reply:
I agree with serializing options in the controller and using the convention `<fieldname>_options`  for now, that is, on the model of `language_options` in the above mentioned example https://github.com/metabrainz/musicbrainz-server/commit/6bf052f9e07bc0125130410a409dcff7924a0638#diff-5053a22c07db9b62eceff446a9b99731R414

Timezone options are serialized the same way at https://github.com/metabrainz/musicbrainz-server/pull/718/commits/e43cd919655f2876c267196503ac28ba2f557214#diff-5053a22c07db9b62eceff446a9b99731R484